### DISCUSSION TITLE
Custom runner for action deployment

### DISCUSSION THREAD

**Discussion by ruchi1803** [community/community#136540] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi Team,

I am new to github actions and trying to explore the automation for deployment. My existing deployment is running in Jenkins. Though I have a lot of questions, but right now I am stuck on one particular one. 

How can I create self hosted agent with my own private docker image. 

This is what all I have done. 
Installed ARC on AKS following this link: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/quickstart-for-actions-runner-controller

I have 2 namespaces one  `arc-systems`  where the controller and listener pods are running and `arc-runners` in which self-agent actions pod get created while triggering the job from github pipeline. 
But this is all working fine for sample application. 
My requirement is for my project where I need use my own docker image with specific tools, and I am not able to get how I can use the self-hosted agent with this image. 
I checked this document https://github.com/actions/actions-runner-controller/blob/master/docs/choosing-runner-destination.md and tried creating the RunnerDeployment I am getting error : 
```no matches for kind "RunnerDeplyment" in version "actions.summerwind.dev/v1alpha1" ensure CRDs are installed first```

 Also as it says warning that this is old way of ARC. Can someone help me what I need to create to use self-hosted agent with private image.

**Comment by danialtavakoli** [comment]
To create a self-hosted GitHub Actions runner using your private Docker image in AKS with Actions Runner Controller, first ensure the CRDs are installed by applying the latest ARC YAML from GitHub. Then, define a `RunnerDeployment` YAML file specifying your custom Docker image and any necessary Docker credentials. Apply this configuration to your AKS cluster. This setup will enable your self-hosted runner to use the custom Docker image. Make sure your CRDs are up-to-date to avoid issues with deprecated methods.

**Comment by ruchi1803** [comment]
Thanks Danial
I have these CRD installed.
gha-runner-scale-set-controller-0.9.0
gha-runner-scale-set-0.9.0

but I can see there is one more in the git repo: `actions-runner-controller`.
I didn't install this one, thought this is old method of using ARC. 
Do I need to install this CRD.

On the other hand I thought if I will use my own image in the gha-runner-scale-set-controller helm values file. like below:

```template:
  spec:
    containers:
      - name: runner
        image: myregistry/dotnet-6-runner:latest```
        
        
 What will be the best way of doing it, updating helm values or creating RunnerDeployment.

**Reply by danialtavakoli** [reply, marked as answer]
You don’t need to install the old `actions-runner-controller` CRD if you’re already using the `gha-runner-scale-set-controller`. The best way to use your custom Docker image is to update the Helm values file for the `gha-runner-scale-set-controller`. Modify the `containers` section like this:

```yaml
spec:
  containers:
    - name: runner
      image: myregistry/dotnet-6-runner:latest
```

This approach directly integrates your custom image into the controller’s configuration, ensuring all runners use it without needing a separate `RunnerDeployment` setup. If this helps, feel free to mark this response as the answer for support.

**Comment by crohr** [comment]
If all you need is a custom image for your runner with preinstalled tooling, you could investigate [RunsOn](https://runs-on.com), which allows bringing your own AMIs, and has the advantage of not having to deal with ARC.

**Comment by armyost** [comment]
You can prebuild some tools inside the runner by yourself.

In my case, I built my runner image and push to private artifact registry.

You can refer the Dockerfile the following link.
https://github.com/actions/runner/blob/main/images/Dockerfile

**Comment by ruchi1803** [comment]
@danialtavakoli I updated my helm chart, but the problem now is while triggering the job none of the agent pods are getting created.
I am assuming there might some customisation that needs to be done in dockerfile. 
There is no error logs in action-controller and listener pod.
